### DISCUSSION TITLE
YAML deprecation duration policy

### DISCUSSION THREAD

**Discussion by frenck** [home-assistant/architecture#920] [General] [answered]
## Context

Currently, when a YAML integration configuration option (or options) gets deprecated, we generally take into account a minimal deprecation period of at least 2 months (being 2 release cycles). After the deprecation period has passed, we allow the YAML configuration option (or options) to be removed.

This is not a formal rule but has been the reviewing standard for years now. 

Some exceptions may occur right now; for example, if an integration is extremely popular (like recently with MQTT), we sometimes use a longer period. Or, we sometimes allow shorter, in case an automated migration path isn't possible.

The consequence of the current deprecation period is that a user should not skip 2 release cycles. If they do, they will miss out on a migration path and thus hit a breaking change, which is not a good user experience.

## Proposal

The proposal is to create an ADR to document a YAML configuration deprecation policy, focusing on minimizing users running into breaking changes by giving them more time to upgrade. To achieve this, the proposal is to extend and formalize the official deprecation period for deprecated YAML configuration options to 6 months (being 6 release cycles at the time of writing).

The 6-months allows a user to upgrade just twice a year to avoid hitting breaking changes, which I think can be assumed as a reasonable minimum upgrade policy to take into account.

Additionally, we set out other requirements for deprecations:

- The use of deprecated YAML configuration must raise an issue in the users' repairs dashboard. This issue should explain why this happens and what a user has to do to resolve the issue. In case an migration is available, this message should only be shown after a successful migration.
- If a migration (e.g., using an import) is possible, it becomes required. If an import has failed, this must be reflected by raising a repair issue on what happened and how to resolve it.

## Consequences

- We need to keep around imports and deprecated YAML configuration options around longer.
- Risks of a user running into a breaking change by skipping a couple of releases will be reduced.

**Comment by balloob** [comment, marked as answer]
5 thumbs up from core members. Approved ✅

**Comment by frenck** [comment]
Opened up PR to formalize this in an ADR: <https://github.com/home-assistant/architecture/pull/922>
